linq4j/src/main/java/org/apache/calcite/linq4j/MergeUnionEnumerator.java

```diff
@@ -0,0 +1,198 @@
+  private final List<Enumerator<TSource>> inputs;
+  private final TSource[] currentInputsValues;
+  private final Function1<TSource, TKey> sortKeySelector;
+  private final Comparator<TKey> sortComparator;
+  private TSource currentValue;
+  private int activeInputs;
+
+  // Set to control duplicates, only used if "all" is false
+  private final @Nullable Set<EnumerableDefaults.Wrapped<TSource>> processed;
+  private final @Nullable Function1<TSource, EnumerableDefaults.Wrapped<TSource>> wrapper;
+
+  private static final Object NOT_INIT = new Object();
+  private static final Object FINISHED = new Object();
+
+  MergeUnionEnumerator(
+      List<Enumerable<TSource>> sources,
+      Function1<TSource, TKey> sortKeySelector,
+      Comparator<TKey> sortComparator,
+      boolean all,
+      EqualityComparer<TSource> equalityComparer) {
+    this.sortKeySelector = sortKeySelector;
+    this.sortComparator = sortComparator;
+
+    if (all) {
+      this.processed = null;
+      this.wrapper = null;
+    } else {
+      this.processed = new HashSet<>();
+      this.wrapper = EnumerableDefaults.wrapperFor(equalityComparer);
+    }
+
+    final int size = sources.size();
+    this.inputs = new ArrayList<>(size);
+    for (Enumerable<TSource> source : sources) {
+      this.inputs.add(source.enumerator());
+    }
+
+    //noinspection unchecked
+    this.currentInputsValues = (TSource[]) new Object[size];
+    this.activeInputs = this.currentInputsValues.length;
+    //noinspection unchecked
+    this.currentValue = (TSource) NOT_INIT;
+
+    initEnumerators();
+  }
+
+  @RequiresNonNull("inputs")
+  @SuppressWarnings("method.invocation.invalid")
+  private void initEnumerators(@UnknownInitialization MergeUnionEnumerator<TSource, TKey> this) {
+    for (int i = 0; i < inputs.size(); i++) {
+      moveEnumerator(i);
+    }
+  }
+
+  private void moveEnumerator(int i) {
+    final Enumerator<TSource> enumerator = inputs.get(i);
+    if (!enumerator.moveNext()) {
+      activeInputs--;
+      //noinspection unchecked
+      currentInputsValues[i] = (TSource) FINISHED;
+    } else {
+      currentInputsValues[i] = enumerator.current();
+    }
+  }
+
+  private boolean checkNotDuplicated(TSource value) {
+    if (processed == null) {
+      return true; // UNION ALL: no need to check duplicates
+    }
+
+    // check duplicates
+    @SuppressWarnings("dereference.of.nullable")
+    final EnumerableDefaults.Wrapped<TSource> wrapped = wrapper.apply(value);
+    if (!processed.contains(wrapped)) {
+      if (!processed.isEmpty()) {
+        // Since inputs are sorted, we do not need to keep in the set all the items that we
+        // have previously returned, just the ones with the same key, as soon as we see a new
+        // key, we can clear the set containing the items belonging to the previous key
+        final EnumerableDefaults.Wrapped<TSource> wrappedItemInSet = processed.iterator().next();
```

Comment:
I guess `iterator().next()` would have noticeable overhead since it is invoked for every row.

Reply:
Changed: keep `currentKeyInProcessedSet` variable